File: app/models/snapshot.py
from app.extensions import db
from datetime import datetime, timezone


class Snapshot(db.Model):
    __tablename__ = "snapshots"

    id = db.Column(db.Integer, primary_key=True)
    tenant_id = db.Column(db.Integer, db.ForeignKey("tenants.id"))
    snapshot_date = db.Column(db.Date, nullable=False)
    uploaded_by = db.Column(db.Integer, db.ForeignKey("users.id"))
    uploaded_at = db.Column(db.DateTime, default=lambda: datetime.now(timezone.utc))
    notes = db.Column(db.Text)

    __table_args__ = (
        db.UniqueConstraint("tenant_id", "snapshot_date", name="uq_tenant_snapshot_date"),
    )

    uploader = db.relationship("User", backref="uploads")
    department_summaries = db.relationship("DepartmentSummary", backref="snapshot", cascade="all, delete-orphan")
    gift_type_breakdowns = db.relationship("GiftTypeBreakdown", backref="snapshot", cascade="all, delete-orphan")
    source_breakdowns = db.relationship("SourceBreakdown", backref="snapshot", cascade="all, delete-orphan")
    fund_breakdowns = db.relationship("FundBreakdown", backref="snapshot", cascade="all, delete-orphan")
    raw_gifts = db.relationship("RawGift", backref="snapshot", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Snapshot {self.snapshot_date}>"

File: src/models/snapshot.js
const { DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  return sequelize.define('Snapshot', {
    id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
    tenantId: { type: DataTypes.INTEGER, field: 'tenant_id' },
    snapshotDate: { type: DataTypes.DATEONLY, allowNull: false, field: 'snapshot_date' },
    uploadedBy: { type: DataTypes.INTEGER, field: 'uploaded_by' },
    uploadedAt: { type: DataTypes.DATE, defaultValue: DataTypes.NOW, field: 'uploaded_at' },
    notes: { type: DataTypes.TEXT },
  }, {
    tableName: 'snapshots',
    timestamps: false,
    indexes: [
      { unique: true, fields: ['tenant_id', 'snapshot_date'] },
    ],
  });
};
